### ISSUE TITLE
Important accuracy drop in onnx conversion, image classification: yolov8n-cls.pt vs yolov8n-cls.onnx (46% similarity)

### ISSUE THREAD

**Issue by xalbertoisorna** [ultralytics/ultralytics#5016]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Export

### Bug

The problem is when testing yolov8-**image-classification** compared to the exported onnx-**image-classification** with the same image size and default parameters, it has a significant drop in precision, only 46% match on 100 COCO train images. 

Tested with 100 coco images and compared the number of matches between the two models. 

The code can be found below, and all images attached to reproduce the issue. 

Is this normal? Why such a big difference?

### Environment

Ultralytics YOLOv8.0.183  Python-3.10.9 torch-2.0.1+cpu CPU (11th Gen Intel Core(TM) i7-1165G7 2.80GHz)
Setup complete  (8 CPUs, 15.7 GB RAM, 325.1/475.9 GB disk)

OS                  Windows-10-10.0.19045-SP0
Environment         Windows
Python              3.10.9
Install             git
RAM                 15.69 GB
CPU                 11th Gen Intel Core(TM) i7-1165G7 2.80GHz
CUDA                None

matplotlib           3.8.0>=3.3.0
numpy                1.26.0>=1.22.2
opencv-python        4.8.0.76>=4.6.0
pillow               10.0.1>=7.1.2
pyyaml               6.0.1>=5.3.1
requests             2.31.0>=2.23.0
scipy                1.11.2>=1.4.1
torch                2.0.1>=1.8.0
torchvision          0.15.2>=0.9.0
tqdm                 4.66.1>=4.64.0
pandas               2.1.1>=1.1.4
seaborn              0.12.2>=0.11.0
psutil               5.9.5
py-cpuinfo           9.0.0

### Minimal Reproducible Example

```python
import numpy as np
import onnxruntime
import yaml

from PIL import Image
from ultralytics import YOLO

HEIGHT, WIDTH  = 224, 224

# remember: pip install ultralytics onnxruntime onnx

# Yolo global
model = YOLO('yolov8n-cls.pt')  # load an official model

# Onnx global
model.export(format="onnx")
session = onnxruntime.InferenceSession('yolov8n-cls.onnx', providers=['CPUExecutionProvider']) # change provider to meet your needs

# Functions
def get_name(arg):
    with open("./metadata.yaml", 'r') as file:
        data = yaml.safe_load(file)
        
    name = data.get('names').get(arg)
    return name

def get_image(img):
    image = Image.open(img)
    image = image.resize((WIDTH,HEIGHT))
    image = np.asarray(image)
    return image

# Runners
def run_yolo(img):
    global model
    results = model.predict(img, verbose=False)  # predict on an image
    arg = results[0].probs.top1
    return get_name(arg)
    
def run_onnx(img):
    global session
    image = get_image(img)
    (input_details, ) = session.get_inputs()
    (output_details, ) = session.get_outputs()
    image = (image / 255.0).astype('float32').transpose(2, 0, 1)
    image = np.expand_dims(image, axis=0)    
    result = session.run([output_details.name], {input_details.name: image})
    arg=int(np.argmax(np.array(result).squeeze(), axis=0))
    return get_name(arg)



if __name__ == "__main__":
    
    import glob
    imgs = glob.glob("image_samples/*.jpg")
    
    matches = 0 
    for img in imgs:
        yolo = run_yolo(img)
        onnx = run_onnx(img)
        print(f"{yolo},{onnx}")
        
        if yolo == onnx:
            matches += 1

    # results
    print(f"\n -> Matches: {matches}/{len(imgs)}")

``` 

Then the output shows:

```shell
 -> Matches: 46/100 👎 
``` 
[demo_onnx_yolo.zip](https://github.com/ultralytics/ultralytics/files/12682881/demo_onnx_yolo.zip)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@xalbertoisorna i'm sorry to hear about the accuracy drop you're experiencing when comparing YOLOB8's image classification model with the exported ONNX model. This most certainly isn't expected behavior as the accuracy between the two should be fairly consistent, not showing such a significant difference.

Several factors could be affecting this:

1. **Model Variation**: It's possible that the model variation is not converted properly when switching from PyTorch to ONNX. YOLOv8's design is complex and not all features may translate perfectly between the two platforms. 

2. **Image Preprocessing Changes**: If the manner in which the image is preprocessed before inference changes slightly, i.e., resizing methods, normalization, etc., between the two models, that can considerably affect the output.

3. **Library and Environment Differences**: The different Python and PyTorch versions, or other library versions, could potentially affect model behavior.

It's important to ensure every step besides the model is kept exactly the same when comparing performances. I'd suggest to verify that all preprocessing methods are identical between the PyTorch and ONNX workflows and verify that the correct versions of all packages are installed.

Unfortunately, without direct access to your setup, it's quite tough for us to understand what exactly is causing this large accuracy drop. I would recommend carefully revisiting your setup and conversions to ensure that everything is being processed correctly. Please don't hesitate to let us know how it goes and whether you still face this issue.

**Comment by xalbertoisorna** [comment]
> @xalbertoisorna i'm sorry to hear about the accuracy drop you're experiencing when comparing YOLOB8's image classification model with the exported ONNX model. This most certainly isn't expected behavior as the accuracy between the two should be fairly consistent, not showing such a significant difference.
> 
> Several factors could be affecting this:
> 
> 1. **Model Variation**: It's possible that the model variation is not converted properly when switching from PyTorch to ONNX. YOLOv8's design is complex and not all features may translate perfectly between the two platforms.
> 2. **Image Preprocessing Changes**: If the manner in which the image is preprocessed before inference changes slightly, i.e., resizing methods, normalization, etc., between the two models, that can considerably affect the output.
> 3. **Library and Environment Differences**: The different Python and PyTorch versions, or other library versions, could potentially affect model behavior.
> 
> It's important to ensure every step besides the model is kept exactly the same when comparing performances. I'd suggest to verify that all preprocessing methods are identical between the PyTorch and ONNX workflows and verify that the correct versions of all packages are installed.
> 
> Unfortunately, without direct access to your setup, it's quite tough for us to understand what exactly is causing this large accuracy drop. I would recommend carefully revisiting your setup and conversions to ensure that everything is being processed correctly. Please don't hesitate to let us know how it goes and whether you still face this issue.

Thanks for your response, I'll take a look to the preprocessing, please note that I added the code and all sample images in a **zip file below**, I attach them here again just in case. Do not hesitate to test them. 

[demo_onnx_yolo.zip](https://github.com/ultralytics/ultralytics/files/12685827/demo_onnx_yolo.zip)

**Comment by glenn-jocher** [comment]
@xalbertoisorna thank you for providing more details and sharing your supporting files, this is very helpful in understanding your situation better.

In regards to your question about the YOLOv8 image classification preprocessing steps, typically these steps would include:

- Image resizing to match the input dimensions expected by the model.
- Normalization of the image data, usually pixel values are scaled to be between 0 and 1, or sometimes they are standardized to have a mean of 0 and standard deviation of 1.
- In some cases, color space conversion (such as from RGB to BGR), may be performed based on the model's training data.
- Finally, the data is usually converted to the tensor format that the model expects.

Do check through your preprocessing steps carefully once more to ensure they're correctly aligned with the YOLOv8 model expectations. Again, I would encourage you to make sure there aren't any discrepancies in your setup and conversion process that may be affecting the model's performance.

After you have checked and adjusted your setup, please let us know if there is any change in the model's accuracy. We're here to help and your feedback can be valuable in improving YOLOv8.

**Comment by xalbertoisorna** [comment]
> @xalbertoisorna thank you for providing more details and sharing your supporting files, this is very helpful in understanding your situation better.
> 
> In regards to your question about the YOLOv8 image classification preprocessing steps, typically these steps would include:
> 
> * Image resizing to match the input dimensions expected by the model.
> * Normalization of the image data, usually pixel values are scaled to be between 0 and 1, or sometimes they are standardized to have a mean of 0 and standard deviation of 1.
> * In some cases, color space conversion (such as from RGB to BGR), may be performed based on the model's training data.
> * Finally, the data is usually converted to the tensor format that the model expects.
> 
> Do check through your preprocessing steps carefully once more to ensure they're correctly aligned with the YOLOv8 model expectations. Again, I would encourage you to make sure there aren't any discrepancies in your setup and conversion process that may be affecting the model's performance.
> 
> After you have checked and adjusted your setup, please let us know if there is any change in the model's accuracy. We're here to help and your feedback can be valuable in improving YOLOv8.

Hi thanks for your support, it is helpful. 
I double-checked preprocessing stages, here below is the answer to your question.

- Image resizing: ✔️ as you can see in the script both sizes are 224,224 as are pretrained weights. Anyway if you force the export parameter "imgsz" to 224, same result, same matches. 
- Normalization: ✔️ I saw that for yolov8 classification is 0 to 1, and just image / 255. Anyway I tested all the methods that you propose, and the results are worse. 
- **Color Space converison**: 💡 Image from Pillow is RGB. I tried BGR and it is indeed better!! 100% score. So, there are 2 things here. Yolov8 should indicate that the image is **BGR** because people not using OpenCV can be confusing. Also, the model is very color dependant, that is not good, but is another topic. 

Would it be possible to have an RGB model, it is something that you consider? In embedded applications, images are usually RGB.

**Comment by glenn-jocher** [comment]
Hi @xalbertoisorna,

We're glad to hear that you found the root of the discrepancy! Agreeing with you, the color space can be a non-intuitive point of consideration. Changing the color space from RGB to BGR is a common preprocessing step placed in consideration in some model architectures, like YOLOv8, as some datasets may have been trained in that particular manner.

To address the observed color dependency of the model, it's important to understand that model performances are largely dependent on how they were trained – including on what kind of data and preprocessing steps used. If the training data was in BGR format, the model will perform best when given images in the same format.

We greatly appreciate your feedback regarding the need for appropriate documentation about the BGR image use with YOLOv8. It's a valid point and we'll certainly take note of it.

In terms of the possibility of an RGB model for embedded applications, it's a valuable suggestion. Having a model variant that accepts RGB inputs directly would indeed be beneficial for certain applications. While we can't guarantee any immediate changes, please know that it's now on our radar and we will consider it as we plan our future model improvements. 

We hope that this clarifies things for you, and thank you again for raising this issue. We're always looking for ways to improve and your feedback definitely helps with that.
